### PR TITLE
Switch CI to GHC 8.10.2

### PR DESCRIPTION
**Pull request by mrBliss**
Update the dependencies to versions compatible with GHC 8.10.2.

Switch over CI to GHC 8.10.2.

### CODE REVIEW

**Comment by mrBliss** [comment]
Draft because it includes #2539 until it is merged and because it relies on unmerged revisions of dependencies.

**Comment by mrBliss** [comment]
This PR is out of date. #2542 has been merged. Once #2547 and/or #2539 are merged, I'll update/rebase this PR.

**Comment by coot** [comment]
Is this pr stale?

**Comment by mrBliss** [comment]
I have updated this PR

File: ouroboros-consensus-test/test-consensus/Test/Consensus/HardFork/Combinator/A.hs [hunk 1]
```diff
@@ -542,7 +542,7 @@ instance SerialiseNodeToClient BlockA Void where
   decodeNodeToClient _ _ = fail "no ApplyTxErr to be decoded"
 
 instance SerialiseNodeToClient BlockA (SomeBlock Query BlockA) where
-  encodeNodeToClient _ _ (SomeBlock q) = case q of {}
+  encodeNodeToClient _ _ = \case {}
```
Comment (coot) [inline ouroboros-consensus-test/test-consensus/Test/Consensus/HardFork/Combinator/A.hs:545]: So Haskell's pattern matching checker got better, right?

File: ouroboros-consensus/ouroboros-consensus.cabal [hunk 2]
```diff
@@ -316,6 +316,11 @@ library
                        -Widentities
                        -Wredundant-constraints
                        -Wmissing-export-lists
+  -- Pattern match checking has improved since 8.6.5. GHC now considers more
+  -- pattern matches as redundant than before, so disable this warning for older
+  -- versions of GHC that still consider them incomplete.
```
Comment (coot) [inline ouroboros-consensus/ouroboros-consensus.cabal:321]: That answers my question,  :).

**Comment by coot** [comment]
`Failed to build Win32-2.8.3.0.` - maybe we should try a newer version? There's `2.10` available on `hackage`.

**Comment by mrBliss** [comment]
> `Failed to build Win32-2.8.3.0.` - maybe we should try a newer version? There's `2.10` available on `hackage`.

Feel free to push a fix to this PR :slightly_smiling_face:

**Comment by coot** [comment]
We need to either wait for `8.10.3` or apply this fix ourselves: https://www.stackage.org/blog/2020/08/ghc-8-10-2-windows-workaround

**Comment by mrBliss** [comment]
I have rebased the PR against master.

**Comment by mrBliss** [comment]
I have again rebased the PR against master.

**Comment by mrBliss** [comment]
@hamishmack Thanks for fixing the Nix issues one by one :slightly_smiling_face:. The [next one](https://hydra.iohk.io/build/4373836/nixlog/1): `cardano-crypto-class-lib-cardano-crypto-class-2.0.0-x86_64-w64-mingw32.drv` fails with:
```
Setup: The program 'pkg-config' version >=0.9.0 is required but it could not
be found.
```

**Comment by hamishmack** [comment]
`The program 'pkg-config' version >=0.9.0 is required but it could not be found` should be fixed by https://github.com/input-output-hk/haskell.nix/pull/874.  Run `niv update haskell.nix` to try it out.

**Comment by mrBliss** [comment]
> `The program 'pkg-config' version >=0.9.0 is required but it could not be found` should be fixed by [input-output-hk/haskell.nix#874](https://github.com/input-output-hk/haskell.nix/pull/874). Run `niv update haskell.nix` to try it out.

Done, thanks again.

Next one :grin: 

https://hydra.iohk.io/build/4404880/nixlog/8
```
..
/nix/store/spnnfafb6hw9cdpmf3whb1lhdicd7p6h-x86_64-w64-mingw32-binutils-2.31.1/bin/x86_64-w64-mingw32-ld: cannot find -lmingwex
/nix/store/spnnfafb6hw9cdpmf3whb1lhdicd7p6h-x86_64-w64-mingw32-binutils-2.31.1/bin/x86_64-w64-mingw32-ld: cannot find -lmsvcrt
/nix/store/spnnfafb6hw9cdpmf3whb1lhdicd7p6h-x86_64-w64-mingw32-binutils-2.31.1/bin/x86_64-w64-mingw32-ld: cannot find /nix/store/kiljkg0bh7kc5cp5a0cqkshnabnfh73c-mingw-w64-6.0.0-x86_64-w64-mingw32/lib/crtend.o: Too many open files
collect2: error: ld returned 1 exit status
`x86_64-w64-mingw32-cc' failed in phase `Linker'. (Exit code: 1)
builder for '/nix/store/h982lcbph4cviik9sxf8chpsw2kazlwd-ouroboros-consensus-byron-test-test-test-0.1.0.0-x86_64-w64-mingw32.drv' failed with exit code 1
```

**Comment by hamishmack** [comment]
I think that new issue should be fixed by https://github.com/input-output-hk/haskell.nix/pull/878,
but I am not sure.  Please can you try `niv update haskell.nix -b hkm/binutils-fix` to see if it helps?  If it does
and @angerman is happy with it we can merge that PR into the master of haskell.nix.

**Comment by mrBliss** [comment]
> I think that new issue should be fixed by [input-output-hk/haskell.nix#878](https://github.com/input-output-hk/haskell.nix/pull/878),
> but I am not sure. Please can you try `niv update haskell.nix -b hkm/binutils-fix` to see if it helps? If it does
> and @angerman is happy with it we can merge that PR into the master of haskell.nix.

That seems to fix the issue! CI (except for the Windows build in GitHub Actions, ping @coot) is now green and finally ready for the switch to GHC 8.10.2, hurray! :tada:. Thanks again @hamishmack.

**Comment by coot** [comment]
I'll try to fix the remaining windows error by the end of this week.

**Comment by hamishmack** [comment]
Now that https://github.com/input-output-hk/haskell.nix/pull/878 has been merged we should `niv update haskell.nix -b master` to set it back to the master branch.

**Comment by mrBliss** [comment]
> Now that [input-output-hk/haskell.nix#878](https://github.com/input-output-hk/haskell.nix/pull/878) has been merged we should `niv update haskell.nix -b master` to set it back to the master branch.

Done

**Comment by mrBliss** [comment]
If CI turns green, we will finally be ready to switch to GHC 8.10.2 :tada:

Many thanks again to @hamishmack and @coot for making this possible.

**Comment by coot** [comment]
> Many thanks again to @hamishmack and @coot for making this possible.

Many thanks to you @mrBliss for pushing this forward!

**Comment by mrBliss** [comment]
There is one testsuite that's taking forever in CI: [Cardano:ouroboros-network-pr-2540:x86_64-w64-mingw32.haskellPackages.ouroboros-network-framework.coverageReport.x86_64-linux](https://hydra.iohk.io/build/4481251/nixlog/2) 
```
typed-protocols
  Network.Typed.Protocol.PingPong.Codec.Cbor

...

  Subscription
    Resolve (Sim):                                             OK (0.04s)
      +++ OK, passed 100 tests.
      
      Resolution Result (100 in total):
      32% A success, AAAA error
      22% A error, AAAA success
      19% A and AAAA error
       9% AAAA before A (Resolution Delay)
       8% A before AAAA
       7% AAAA before A
       3% A error, AAAA no result
    Resolve Subscribe (IO):                                    OK (5.88s)
      +++ OK, passed 100 tests.
    Send Recive with Dns worker (IO):                          OK (12.73s)
      +++ OK, passed 100 tests.
```
and then it hangs.

The [next property to test](https://github.com/input-output-hk/ouroboros-network/blob/a09f209bb6bbb1a1a9a5b31d671da0189001e175/ouroboros-network-framework/test/Test/Ouroboros/Network/Subscription.hs#L120-L121) is [`prop_send_recv_init_and_rsp`](https://github.com/input-output-hk/ouroboros-network/blob/a09f209bb6bbb1a1a9a5b31d671da0189001e175/ouroboros-network-framework/test/Test/Ouroboros/Network/Subscription.hs#L649).

This same test has timed out before: https://hydra.iohk.io/build/4447199/nixlog/2 (same branch).

@coot Do you know what could be wrong?

**Comment by mrBliss** [comment]
bors merge

**Review by coot — APPROVED** [review]
LGTM, though majority of changes are in consensus (or nix).

**Comment by mrBliss** [comment]
bors merge

**Comment by mark-stopka** [comment]
Congratulations guys! @mrBliss and @coot for your patience, I hope me and @mimi1vx will be able to contribute more and more... especially if DC fund approves #dCloud forward, I have already spoke with Alex from Ergo about incirporating Ergo NIPoPoW from Ergo to Cardano... then the other way around Ouroboros Praos -> Ergo...

Our goal currently is to make Daedalus available as RPM package in openSUSE and CentOS, possibly also Ubuntu later... + add a ton of integration testing in this new repo...

https://github.com/2nd-Layer/cardano-core-integration-testing

@CharlesHoskinson as we are such a "friends", are there people at IOG I could use as point of contacts as we move forward? If you read the (sadly Czech) article, our partner, Invictus Mining has 330M CZK (14M USD) budget (incl. HW capex) to build either a dCloud or an altrnative to dCloud on other main chain, either way if dCloud won't get funded I would be joining their team do build the storage piece... We both know Decentralized Software Updates & Voltaire will need a storage solution, can we start the conversation about that over here?

https://github.com/input-output-hk/decentralized-software-updates/issues/150

**Pull request merged**